compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -271,14 +271,18 @@ trait TypeAssigner {
   def assignType(tree: untpd.Select, qual: Tree)(implicit ctx: Context): Select = {
     def qualType = qual.tpe.widen
     def arrayElemType = {
-      val JavaArrayType(elemtp) = qualType
-      elemtp
+      qualType match {
+        case JavaArrayType(elemtp) => elemtp
+        case _ =>
+          ctx.error("Array type conflict with " + qual.symbol.name, tree.pos)
+          defn.NothingType
+      }
```

Comment:
Scalac allows both the following:

```Scala
class Foo[U] { self : Array[U] => self(0) }

class Foo[U] { self : Array[U] with Nothing =>
  self.length
}
```

We need to justify why Dotty should deviate in such cases.

Reply:
We will never be able to instantiate a Foo because Array is final. The code is nonsensical as no subclass of Foo can mixin Array.

Reply:
I'm thinking it's the same case as `def f(x: Nothing): Unit = 3`.  The code is valid, though the function can not be called. If we have checks for realization of self types during instantiation, can we just reuse that mechanism?

Reply:
Maybe we could. But this patch should still be there to ensure there are no compiler crashes on arrays in case that checker fails to detect it early.

Also, It is possible that we may create some other kind of unrealisable array with something that is not the self type.

Reply:
I tend to agree with @liufengyun. We should fix https://github.com/lampepfl/dotty/blob/4c8a2bd319437f7ab22303a162f5922bc000134d/compiler/src/dotty/tools/dotc/typer/Checking.scala#L148
Otherwise we will reject `class Foo[U] { self : Array[U] & Nothing => self(0) }` but accept `class Foo[U] { self : Array[U] & Nothing => }`.

> Also, It is possible that we may create some other kind of unrealisable array with something that is not the self type.

Wouldn't that be a bug? Then crashing is the right think to do

Reply:
I tried the fix on latest master (with #4248) and the following code snippet compiles:
```scala
class Foo { def test[U](x: Array[U] & Nothing) = x(0) }
```

Reply:
That was a different issue.

Reply:
The `& Nothing` is unrelated tho this particular issue.

Reply:
> That was a different issue

Sorry, my point was that we should also reject ` test[U](x: Array[U] & Nothing) = x(0)`, if we reject `self: Array[U] & Nothing => self(0)`

Reply:
> If we have checks for realization of self types during instantiation, can we just reuse that mechanism?

BTW, those checks on instantiation are spec-mandated (#4252), so I think I agree with @allanrenucci and @liufengyun.

Reply:
The check on instantiation are done and work as expected. The issue is that for this particular case we do have a problem before instantiation. This is a particular exception due to the way we erase Arrays into java arrays. 

Additionally the self type Array is completely valid as far as this class is conserened and could be realizable. Though Array would be the only class that could extend it.